### PR TITLE
notesnook: init at 2.5.7

### PR DESCRIPTION
**Pull request by j0lol**
###### Description of changes

https://notesnook.com/

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by j0lol** [comment]
I tried to build by source, but there is a git dep without a lockfile, which is unsupported by buildNpmPackage. darwin is also untested.

**Comment by thecodrr** [comment]
@j0lol which dep is it? Maybe we can replace/fix it.

**Comment by j0lol** [comment]
@thecodrr 

This dependency is not present in package-lock.json: https://github.com/streetwriters/notesnook/blob/c0bc7866693a8f20eea20fbc9f87a2945c77b648/apps/web/package.json#L48

from nixos discord:
> localforage-getitems. If it's uploaded to the registry it would work fine (since then npm doesn't need to special-case the build process) or it can be replaced, have package-lock.json added, etc.

**Comment by thecodrr** [comment]
@j0lol 

https://github.com/streetwriters/notesnook/blob/c0bc7866693a8f20eea20fbc9f87a2945c77b648/apps/web/package-lock.json#L26058

https://github.com/streetwriters/notesnook/blob/c0bc7866693a8f20eea20fbc9f87a2945c77b648/apps/web/package-lock.json#L12209

**Comment by lilyinstarlight** [comment]
@thecodrr Hi! I'm the person from Discord that @j0lol was quoting. I think it got a little mangled in translation, but specifically https://github.com/thecodrr/localForage-getItems does not have a package-lock.json in its repo, only a yarn.lock

We would like to support that case eventually (since it is still reproducible with a yarn.lock), but for now in `buildNpmPackage` we can only deal with git deps that have a package-lock.json file (it only matters because pacote/npm runs install scripts for git deps, which requires dev dependencies that aren't locked in the downstream dependent package-lock.json, only the upstream one -- for registry deps, install scripts aren't run so it's not a problem there iirc)

So the options that could make it immediately supported for a from-source build via `buildNpmPackage` would be to either upload it to the registry or add a package-lock.json to that repository (since it does appear to be yours)

Thanks for chiming in on this PR and trying to assist :)

**Comment by thecodrr** [comment]
@lilyinstarlight ah, now I get it. Alright, I'll add the package-lock.json file and report back.

**Comment by j0lol** [comment]
odd issue, links don't open firefox. this is a documented issue for discord and im sure other electron apps suffer from it. i tried to override nss but it didn't work.

**Comment by thecodrr** [comment]
@lilyinstarlight I have fixed the issue here. Notesnook now uses npm published `@notesnook/localforage-getitems` package. Let me know if I can help in any other way.

**Comment by lilyinstarlight** [comment]
@thecodrr Looking closer it turns out it's the same deal with https://github.com/thecodrr/node-html-parser in the notesnook-importer packages, but that's the only other problem. It's missing a package-lock.json in the repository and is included as a git dep on a few importer packages

(https://github.com/thecodrr/node-html-to-text is also included as a git dep from `@notesnook/core`, but it does have a package-lock.json file in the repo and so should work with our tooling as-is)

Thank you again!

**Comment by lilyinstarlight** [comment]
@thecodrr Have you had a chance to look at and do the same thing for https://github.com/thecodrr/node-html-parser? Thank you again!

**Comment by thecodrr** [comment]
@lilyinstarlight should be fixed in https://github.com/streetwriters/notesnook/commit/ebc4324ef1651f4371c59d18785eb58ca03d8031

**Comment by thecodrr** [comment]
@j0lol what's the status for this?

**Comment by j0lol** [comment]
i dont have the version of the package that builds from source, ive been procrastinating looking for it or rewriting it, i should do that soon haha

sorry for taking a while!

**Comment by lilyinstarlight** [comment]
@j0lol, I'm trying to get the pending `fetchNpmDeps`-related and `buildNpmPackage`-related PRs merged in the next few days, so this will need those first

I've copied the work-in-progress derivation I was testing with below. It still needs some work though (also I apologize this all has taken so much time... thank you for letting us use it to iron out our tooling bugs though!)

<details>
<summary>notesnook.nix</summary>

```nix
{ lib
, buildNpmPackage
, fetchFromGitHub
, buildPackages
, nodePackages
, pkg-config
, python3
, libsecret
}:

buildNpmPackage rec {
  pname = "notesnook";
  version = "unstable-2023-03-10";

  src = fetchFromGitHub {
    owner = "streetwriters";
    repo = "notesnook";
    rev = "ebc4324ef1651f4371c59d18785eb58ca03d8031";
    hash = "sha256-aQjHLdBdrzdk3UkV/CnxLzaV8hT8NyAANdxEpTKvn8c=";
  };

  sourceRoot = "source/apps/web";

  npmDepsHash = "sha256-fPBYW67TDNMglTbNOpEnHWVo83YDsKHMrBS54MnGWig=";
  npmFlags = [ "--verbose" ];
  makeCacheWritable = true;

  REACT_APP_GIT_HASH = lib.substring 0 9 src.rev;

  nativeBuildInputs = [
    pkg-config
    python3
    nodePackages.patch-package
  ];

  buildInputs = [
    libsecret
  ];

  postPatch = ''
    # deal with lerna workspace packages without lerna
    sed -i package.json \
      -e 's|"@notesnook/desktop":\s*"\*"|"@notesnook/desktop": "file:desktop"|' \
      -e 's|"@notesnook/web-clipper":\s*"\*"|"@notesnook/web-clipper": "file:../../packages/clipper"|' \
      -e 's|"@notesnook/\([^"]*\)":\s*"\*"|"@notesnook/\1": "file:../../packages/\1"|'
    HOME=$TMPDIR ${lib.getBin buildPackages.nodejs}/bin/npm install --package-lock-only --ignore-scripts

    # avoid npx call and use command from native build input
    substituteInPlace ../../packages/editor/package.json --replace 'npx patch-package' 'patch-package'

    # no git hash in build dir so inject one
    substituteInPlace .env-cmdrc.js --replace 'execSync("git rev-parse --short HEAD").toString().trim()' 'process.env.REACT_APP_GIT_HASH'
  '';

  meta = with lib; {
    description = "A fully open source & end-to-end encrypted note taking alternative to Evernote";
    longDescription = ''
      Notesnook is a free (as in speech) & open source note taking app
      focused on user privacy & ease of use. To ensure zero knowledge
      principles, Notesnook encrypts everything on your device using
      XChaCha20-Poly1305 & Argon2.
    '';
    homepage = "https://notesnook.com";
    license = licenses.gpl3Plus;
    maintainers = with maintainers; [ j0lol ];
  };
}
```
</details>

**Comment by thecodrr** [comment]
@lilyinstarlight no worries. Let me know if you need any help regarding this.

File: pkgs/applications/misc/notesnook/default.nix [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+      XChaCha20-Poly1305 & Argon2.
+    '';
+    homepage = "https://notesnook.com";
+    license = licenses.gpl3Only;
+    maintainers = with maintainers; [ j0lol ];
+    platforms = [ "x86_64-linux" "x86_64-darwin" "aarch64-darwin" ];
+  };
+
+  linux = appimageTools.wrapType2 rec {
+    inherit name src meta;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/notesnook/default.nix:45]:
```suggestion
    inherit pname src meta;
```

File: pkgs/applications/misc/notesnook/default.nix [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+      mv $out/bin/{${name},${pname}}
+      install -Dm444 ${appimageContents}/notesnook.desktop -t $out/share/applications
+      install -Dm444 ${appimageContents}/notesnook.png -t $out/share/pixmaps
+      substituteInPlace $out/share/applications/notesnook.desktop \
+        --replace 'Exec=AppRun --no-sandbox %U' 'Exec=${pname}'
+    '';
+  };
+
+  darwin = stdenv.mkDerivation {
+    inherit name src meta;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/notesnook/default.nix:63]:
```suggestion
    inherit pname version src meta;
```

File: pkgs/applications/misc/notesnook/default.nix [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    url = "https://github.com/streetwriters/notesnook/releases/download/v${version}/notesnook_${suffix}";
+    hash = {
+      x86_64-linux = "sha256-L+nVTqT8CVwrslhiF5xj77CnrpW6c17rK+g540T4BT0=";
+      x86_64-darwin = "sha256-nbE2G1d2fpbBS756T3b4L03xP8JlCR5Ez6FVANWNHK0=";
+      aarch64-darwin = "sha256-ZVi0t5LmmUiHrrD58nllLTGrSTIJIcNNYIzCkHYVGoo=";
+    }.${system} or throwSystem;
+  };
+
+  appimageContents = appimageTools.extractType2 {
+    inherit name src;
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/notesnook/default.nix:27]:
```suggestion
    inherit pname version src;
```

File: pkgs/applications/misc/notesnook/default.nix [hunk 4]
```diff
@@ -0,0 +1,77 @@
+{ lib, stdenv, appimageTools, fetchurl, undmg }:
+
+let
+  pname = "notesnook";
+  version = "2.3.0";
+  name = "${pname}-${version}";
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/notesnook/default.nix:6]:
```suggestion
```

**Review by SuperSandro2000 — COMMENTED** [review]
all app image tools support now pname version.

**Comment by thecodrr** [comment]
@lilyinstarlight @j0lol what's the status on this?

**Comment by j0lol** [comment]
I'm personally maintaining the appimage version of this package on my personal package repo (`notesnook-appimage`): https://codeberg.org/j0/jopkgs. I also have a work-in-progress from-source package there (`notesnook`, mostly from the package from lily above), but I'm not sure how to make it build yet. 

Ideally the package that gets into nixpkgs should be building from source instead of repackaging the appimage...

Here's the log:
```
@nix { "action": "setPhase", "phase": "unpackPhase" }
unpacking sources
unpacking source archive /nix/store/wrd0mi4acydqy27kgwwy38iw08m7k6xl-source
source root is source/apps/web
@nix { "action": "setPhase", "phase": "patchPhase" }
patching sources
[..................] \ idealTree: timing arborist:ctor Completed in 1ms^M[..................] - idealTree:web: sill idealTree buildDeps^M[#########.........] - idealTree:desktop: timing idealTree:node_modules/@ <SNIP>
npm ERR! syscall getaddrinfo
npm ERR! errno EAI_AGAIN
npm ERR! request to https://registry.npmjs.org/diary failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org

npm ERR! A complete log of this run can be found in:
npm ERR!     /build/.npm/_logs/2023-06-07T17_08_08_280Z-debug-0.log
/nix/store/0swa08bbslxc9wbfixbcd196aa1117lx-stdenv-linux/setup: line 134: pop_var_context: head of shell_variables not a function context
```
The "complete log of this run" is completely empty.

**Comment by j0lol** [comment]
Going to bump & submit this without from-source for now. Hopefully we can replace this with a from-source build when possible :)

**Comment by thecodrr** [comment]
> npm ERR! request to https://registry.npmjs.org/diary failed, reason: getaddrinfo EAI_AGAIN registry.npmjs.org

This seems like a Network issue.

**Comment by j0lol** [comment]
@lilyinstarlight could we get this merged? :)

**Comment by SuperSandro2000** [comment]
> all app image tools support now pname version.

you reverted by previous review comments

**Comment by mannp** [comment]
Looking forward to using my notesnook pro subscription with this lovely app, once the team get chance to merge it :) TIA

File: pkgs/applications/misc/notesnook/default.nix [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    inherit name src meta;
+
+    profile = ''
+      export LC_ALL=C.UTF-8
+    '';
+
+    multiPkgs = null; # no 32bit needed
+    extraPkgs = appimageTools.defaultFhsEnvArgs.multiPkgs;
+    extraInstallCommands = ''
+      mv $out/bin/{${name},${pname}}
```
Comment (SuperSandro2000) [inline pkgs/applications/misc/notesnook/default.nix:54]:
```suggestion
      mv $out/bin/{${pname}-${version},${pname}}
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]
Please take a look at the review comments

**Comment by j0lol** [comment]
whoops! sorry about that, resolved

**Comment by j0lol** [comment]
@thecodrr don't think we have a good way to use lerna in nix yet... @lilyinstarlight was working on that part of the package but i haven't heard back in a while.

**Comment by lilyinstarlight** [comment]
> I am not sure why you are doing it this way. Is there a problem with using npx/lerna in nix? Also since this is a monorepo, the packages are heavily interlinked so you'd have to do the above substitution in many places in order for this to work.

Yeah that was a bit of a hacky thing I was doing for testing stuff. We cannot use lerna exactly with our tooling since we only support npm right now and lerna assumes it has internet access (which is not the case in the sandbox, to ensure reproducibility)

> I actually got rid of Lerna in the latest master branch substituting it for using a custom node script. Will that work?

Maybe! I'll take a look sometime in the next few (maybe Tuesday) and report back after some experimentation

We're going to go ahead and merge the binary package for now once @SuperSandro2000's review items have all been addressed (the `pname`/`version` ones are still not addressed) and I'm going to look at working on our tooling enough to make the current state of the notesnook repository work reproducibly with it

Thank you for your patience and cooperation!

**Comment by j0lol** [comment]
sorry for all the mess with that review, been working on this for a while haha, all good now

**Review by lilyinstarlight — APPROVED** [review]
This looks good enough to me pending ofborg builds succeeding. Do you have any more comments @SuperSandro2000 or is this good for merge?

**Comment by SuperSandro2000** [comment]
@ofborg eval

**Pull request merged**